### DISCUSSION TITLE
IKEv1 - about "inactivity = <time>" and where child SAs take that time value

### DISCUSSION THREAD

**Discussion by SP-LukasG** [strongswan/strongswan#1564] [General]
Hello to everyone,

we are struggling with duplicating SAs mostly between strongswan<->strongswan S2S connections.
When i say struggling i mean not a few, but hundreds to thousands upon open, unused SAs which are constantly rekeying. 
Because the duplicated child SAs are not processing any traffic, I thought the option "inactivity" would be perfect.
But i observed an interesting phenomenon with the "in" and "out" timings of child SAs.

Here is an example of the config:

```
connections {
    IPSec_200$2E4$2E200$2E17 {
        version = 1           
        local_addrs = 100.4.159.246
        remote_addrs = 200.4.200.17
        local {                    
            auth = psk
            id = 100.4.159.246
        }                     
        remote {
            auth = psk
            id = 200.4.200.17
            revocation = strict
        }                      
        reauth_time=51m
        over_time=9m   
        rand_time=9m
        keyingtries=0
        proposals = aes256-sha2_512-ecp256
        children {                        
            IPSec_200$2E4$2E200$2E17_12 {
                start_action = trap      
                dpd_action = trap  
                ipcomp = no      
                remote_ts = 192.168.143.0/24
                local_ts = 192.168.146.0/24 
                mode = tunnel              
                life_time=60m
                rekey_time=51m
                rand_time=9m  
                close_action = start
                inactivity=1m                           
                esp_proposals = "aes256-sha2_512-ecp256"    
            }
        }
    }
}   
```                
An example of open SAs would be this:

```
IPSec_200$2E4$2E200$2E17: #1, ESTABLISHED, IKEv1, 13d0b073d5d99e9e_i 9cf5c492fa4e71e0_r*
  local  '100.4.159.246' @ 100.4.159.246[500]
  remote '200.4.200.17' @ 200.4.200.17[500]
  AES_CBC-256/HMAC_SHA2_512_256/PRF_HMAC_SHA2_512/ECP_256
  established 138s ago, reauth in 2584s
  IPSec_200$2E4$2E200$2E17_12: #3, reqid 1, REKEYED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 19s ago, rekeying in 2819s, expires in 3581s
    in  c9eb556f,      0 bytes,     0 packets
    out c062f0fc,      0 bytes,     0 packets
    local  192.168.146.0/24
    remote 192.168.143.0/24
  IPSec_200$2E4$2E200$2E17_12: #5, reqid 1, REKEYED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 18s ago, rekeying in 2984s, expires in 3582s
    in  cd420082,      0 bytes,     0 packets
    out c1ff8933,      0 bytes,     0 packets
    local  192.168.146.0/24
    remote 192.168.143.0/24
  IPSec_200$2E4$2E200$2E17_12: #7, reqid 1, REKEYED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 17s ago, rekeying in 2622s, expires in 3583s
    in  cd904765,      0 bytes,     0 packets
    out c29558e6,      0 bytes,     0 packets
    local  192.168.146.0/24
    remote 192.168.143.0/24
  IPSec_200$2E4$2E200$2E17_12: #9, reqid 1, INSTALLED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 15s ago, rekeying in 2666s, expires in 3585s
    in  c1ca985f,      0 bytes,     0 packets
    out cbeb2d11,      0 bytes,     0 packets
    local  192.168.146.0/24
    remote 192.168.143.0/24
  IPSec_200$2E4$2E200$2E17_12: #10, reqid 1, INSTALLED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 15s ago, rekeying in 2826s, expires in 3585s
    in  cdf81a2a,     84 bytes,     4 packets,     8s ago
    out ce90b28e,     84 bytes,     6 packets,     7s ago

```
My assumption was, that the unused SAs will be closed 60 seconds after they were installed and no packets were transmitted.
What happens in stead is, that these SAs inherit the "in" and "out" traffic timers, after 60 seconds, from the only SA which actually is in use. 

This is an example of that:

```
IPSec_100$2E4$2E200$2E17: #1, ESTABLISHED, IKEv1, 13d0b073d5d99e9e_i 9cf5c492fa4e71e0_r*
  local  '100.4.159.246' @ 100.4.159.246[500]
  remote '200.4.200.17' @ 200.4.200.17[500]
  AES_CBC-256/HMAC_SHA2_512_256/PRF_HMAC_SHA2_512/ECP_256
  established 581s ago, reauth in 2141s
  IPSec_200$2E4$2E200$2E17_12: #3, reqid 1, REKEYED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 462s ago, rekeying in 2376s, expires in 3138s
    in  c9eb556f,      0 bytes,     0 packets,    11s ago
    out c062f0fc,      0 bytes,     0 packets,    10s ago
    local  192.168.146.0/24
    remote 192.168.143.0/24
  IPSec_200$2E4$2E200$2E17_12: #5, reqid 1, REKEYED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 461s ago, rekeying in 2541s, expires in 3139s
    in  cd420082,      0 bytes,     0 packets,     1s ago
    out c1ff8933,      0 bytes,     0 packets,     9s ago
    local  192.168.146.0/24
    remote 192.168.143.0/24
  IPSec_200$2E4$2E200$2E17_12: #7, reqid 1, REKEYED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 460s ago, rekeying in 2179s, expires in 3140s
    in  cd904765,      0 bytes,     0 packets,    41s ago
    out c29558e6,      0 bytes,     0 packets,    50s ago
    local  192.168.146.0/24
    remote 192.168.143.0/24
  IPSec_200$2E4$2E200$2E17_12: #9, reqid 1, INSTALLED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 458s ago, rekeying in 2223s, expires in 3142s
    in  c1ca985f,      0 bytes,     0 packets,    40s ago
    out cbeb2d11,      0 bytes,     0 packets,    49s ago
    local  192.168.146.0/24
    remote 192.168.143.0/24
  IPSec_200$2E4$2E200$2E17_12: #10, reqid 1, INSTALLED, TUNNEL, ESP:AES_CBC-256/HMAC_SHA2_512_256/ECP_256
    installed 458s ago, rekeying in 2383s, expires in 3142s
    in  cdf81a2a,   3864 bytes,    46 packets,     1s ago
    out ce90b28e,   3780 bytes,    45 packets,     9s ago
    local  192.168.146.0/24
    remote 192.168.143.0/24

```
What you can see here is, that the SA " IPSec_200$2E4$2E200$2E17_12: #5" reached the 60 seconds limit and got the traffic timings from the SA "PSec_200$2E4$2E200$2E17_12: #10".
From here, this will count up till "in" or "out" reaches 60 seconds and the timings will be replaced again and again. 
For every child SA.

My questions are:
- Why does it behave like this?
- Are there other ways to terminate duplicate child SAs?
- Am I misunderstanding the "inactivity" option?

**Comment by tobiasbrunner** [comment]
> we are struggling with duplicating SAs mostly between strongswan<->strongswan S2S connections.

Please never ever use IKEv1 between two strongSwan instances!

> Why does it behave like this?

Need to read the logs to see what's actually happening. But using `close_action=restart` together with `start_action = trap` in older versions (before 5.9.6, what version are you using?) is definitely a way to create duplicate CHILD_SAs. If the peer closes an SA for some reason, it gets deleted and restarted. But if there is traffic at the same time that triggers the trap policy, another CHILD_SA will be queued.

> Are there other ways to terminate duplicate child SAs?

With 5.9.6+ no duplicates are initiated in the above scenario. Also note that there is an option to delete rekeyed IKEv1 CHILD_SA via `charon.delete_rekeyed` option. But again, don't use IKEv1!

> Am I misunderstanding the "inactivity" option?

No, but before a recent kernel patch and strongSwan change (requires a 6.2 kernel), the last use time had to be retrieved from the policies. And those apply to all duplicate CHILD_SAs.

**Reply by SP-LukasG** [reply]
Thanks for the reply, sorry that I forgot some infos:
Strongswan-Version: 5.9.7
Linux-Kernel: 5.10.149-yocto-standard

> No, but before a recent kernel patch and strongSwan change (requires a 6.2 kernel), the last use time had to be retrieved from the policies. And those apply to all duplicate CHILD_SAs.

I think this is my preferred way to go from here. I will try this with a newer kernel version and see if this helps with duplicate SAs.
The "inactivity" option should work as expected than. 

Many Thanks.

**Reply by tobiasbrunner** [reply]
> With 5.9.6+ no duplicates are initiated in the above scenario.

Actually, I forgot, that's only the case for IKEv2. So even with 5.9.7 this won't apply for IKEv1.

> I think this is my preferred way to go from here. I will try this with a newer kernel version and see if this helps with duplicate SAs.

You also need the current strongSwan release candidate (5.9.10rc1).
